robustifier/src/main/kotlin/cmu/isr/tolerance/SafetyProp.kt

```diff
@@ -0,0 +1,93 @@
+package cmu.isr.tolerance
+
+import addPerturbations
+import alphabetToSet
+import atLeastAsPowerful
+import cmu.isr.ts.lts.*
+import net.automatalib.util.automata.builders.AutomatonBuilders
+import net.automatalib.words.Alphabet
+import net.automatalib.words.impl.Alphabets
+import product
+import satisfies
+
+fun allPerturbations(states : Collection<Int>, alphabet : Alphabet<String>) : Set<Set<Triple<Int, String, Int>>> {
+    fun pertHelper(perturbations : MutableSet<MutableSet<Triple<Int,String,Int>>>,
+                   powerset : Set<Triple<Int,String,Int>>) {
+        if (powerset.isNotEmpty()) {
+            val elem = powerset.first()
+            pertHelper(perturbations, powerset - elem)
+
+            val dPlusElems = mutableSetOf<MutableSet<Triple<Int, String, Int>>>()
+            for (d in perturbations) {
+                dPlusElems += (d + elem) as MutableSet<Triple<Int, String, Int>>
+            }
+            perturbations += dPlusElems
```

Comment:
This looks problematic. + will return a new MutableSet, it does not add dPlusElems to the original set. Or maybe I'm wrong.

Reply:
But += should mutate the set right?

Reply:
That's the part I'm not sure. If you want to add dPlusElems to the perturbations set that is created outside, you might want to use perturbations.addAll(dPlusElems) which provides a more precise semantic.

Reply:
You make a good point so I ran a quick test:

val s = mutableSetOf(0, 1)
s += mutableSetOf(1, 2, 3)
println("Set: {${s.joinToString()}}")

prints out:
Set: {0, 1, 2, 3}

So it looks like += does in fact mutate the set. I also assume this is the right documentation for the += operator: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/plus-assign.html

Reply:
I see, they seem to override that operator.

Reply:
No, += is more like a syntax sugar in Kotlin. You can override operators in Kotlin but not in Java. And Automatalib is based on pure Java, so it might not have this capability.